include/aws/crt/mqtt/Mqtt5Listener.h

```diff
@@ -0,0 +1,344 @@
+#include <aws/crt/mqtt/Mqtt5Client.h>
+#include <aws/crt/mqtt/Mqtt5Types.h>
+#include <aws/mqtt/v5/mqtt5_listener.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        namespace Mqtt5
+        {
+            class Mqtt5ListenerOptions;
+
+            /**
+             * Type signature of the callback invoked when a PacketPublish message received for Mqtt5ListenerCore.
+             *
+             * @return true if the message get handled otherwise return false
+             */
+            using OnListenerPublishReceivedHandler = std::function<bool(const PublishReceivedEventData &)>;
+
+            /**
+             * An MQTT5 listener core. The core is only internally used by Mqtt5Listener to manage the callbacks and
+             * underlying c objects.
+             *
+             * This is a move-only type. Unless otherwise specified, all function arguments need only to live through
+             * the duration of the function call.
+             */
+            class Mqtt5ListenerCore final : public std::enable_shared_from_this<Mqtt5ListenerCore>
+            {
+              public:
+                /**
+                 * Factory function for mqtt5 listener
+                 *
+                 * @param options: Mqtt5ListenerOption
+                 * @param client: Mqtt5Client to listen to
+                 * @param allocator allocator to use
+                 * @return a new mqtt5 listener core
+                 */
+                static std::shared_ptr<Mqtt5ListenerCore> NewMqtt5ListenerCore(
+                    const Mqtt5ListenerOptions &options,
+                    const std::shared_ptr<Mqtt5Client> &client,
+                    Allocator *allocator = ApiAllocator()) noexcept;
+
+                /**
+                 * Get shared pointer of the Mqtt5ListenerCore. Mqtt5ListenerCore is inherited to
+                 * enable_shared_from_this to help with memory safety.
+                 *
+                 * @return shared_ptr for the Mqtt5ListenerCore
+                 */
+                std::shared_ptr<Mqtt5ListenerCore> getptr() { return shared_from_this(); }
+
+                /**
+                 * @return true if the instance is in a valid state, false otherwise.
+                 */
+                operator bool() const noexcept;
+
+                /**
+                 * @return the value of the last aws error encountered by operations on this instance.
+                 */
+                int LastError() const noexcept;
+
+                /*
+                 * Insert the topic and callback in the map
+                 */
+                void Subscribe(Crt::String topic, Crt::Mqtt5::OnPublishReceivedHandler callback) noexcept;
+
+                /*
+                 * Remove the topic and callback in the map
+                 */
+                void Unsubscribe(Crt::String) noexcept;
+
+                /**
+                 * Tells the listener to release the native listener and clean up unhandled the resources
```

Comment:
trivial
"Tells the listener to release the native listener and clean up unhandled resources and operations before destroying it. You MUST only call this function when you want to release the listener.
This is "an ugly and unfortunate necessity" before releasing the Mqtt5ListenerCore. You MUST call this function to avoid memory leaks and/or dead locks."